### PR TITLE
Add option to handle http client redirect

### PR DESCRIPTION
**Pull request by testn**
Signed-off-by: Testo Nakada <test1@doramail.com>

### CODE REVIEW

File: src/main/java/io/vertx/core/http/impl/HttpClientRequestImpl.java [hunk 1]
```diff
@@ -57,12 +60,13 @@
...
   private static final Logger log = LoggerFactory.getLogger(HttpClientRequestImpl.class);
 
-  private final String host;
-  private final int port;
+  private String host;
+  private int port;
+  private int redirectCount;
   private final HttpClientImpl client;
-  private final HttpRequest request;
+  private HttpRequest request;
```
Comment (cescoffier) [inline src/main/java/io/vertx/core/http/impl/HttpClientRequestImpl.java]: if request is not final anymore, access need to be guarded by the monitor lock. It's not the case in uri().

File: src/main/java/io/vertx/core/http/impl/HttpClientRequestImpl.java [hunk 2]
```diff
@@ -57,12 +60,13 @@
...
   private final VertxInternal vertx;
-  private final io.vertx.core.http.HttpMethod method;
+  private io.vertx.core.http.HttpMethod method;
```
Comment (cescoffier) [inline src/main/java/io/vertx/core/http/impl/HttpClientRequestImpl.java]: missing synchronization (same as for request)

File: src/main/java/io/vertx/core/http/impl/HttpClientRequestImpl.java [hunk 3]
```diff
@@ -351,6 +378,42 @@ synchronized void handleResponse(HttpClientResponseImpl resp) {
...
+    }
+
+    resp.endHandler((nothing) -> {
+      String newUri = absoluteUri.toString();
+      reset(newUri);
+      end();
+    });
+  }
+
+  private void reset(String newUri) {
```
Comment (cescoffier) [inline src/main/java/io/vertx/core/http/impl/HttpClientRequestImpl.java]: unprotected access to various non-final fields.

File: src/test/java/io/vertx/test/core/HttpTest.java [hunk 4]
```diff
@@ -4305,6 +4305,138 @@ public void testRequestEnded() {
...
+      });
+      post.end();
+    });
+    await();
+  }
+
+  @Test
   public void testRequestEndedNoEndHandler() {
     server = vertx.createHttpServer(new HttpServerOptions().setPort(HttpTestBase.DEFAULT_HTTP_PORT));
     server.requestHandler(req -> {
```
Comment (purplefox) [inline src/test/java/io/vertx/test/core/HttpTest.java]: Would need to add tests to the HttpClientOptions tests too.

File: src/main/java/io/vertx/core/http/impl/HttpClientRequestImpl.java (hunk 2 above)
Comment (testn) [inline src/main/java/io/vertx/core/http/impl/HttpClientRequestImpl.java]: Need to be able to change that for HTTP 303 redirect where the method can be changed from POST to GET
https://en.wikipedia.org/wiki/HTTP_303

File: src/main/java/io/vertx/core/http/impl/HttpClientRequestImpl.java (hunk 3 above)
Comment (testn) [inline src/main/java/io/vertx/core/http/impl/HttpClientRequestImpl.java]: To handle the redirect seamlessly without creating a new HttpRequest, we need to modify some non-final fields when the redirection happens.

Comment (cescoffier) [inline src/main/java/io/vertx/core/http/impl/HttpClientRequestImpl.java]: Ok, but then all access to the field must be guarded. It's not the case. So there are race conditions.

File: src/main/java/io/vertx/core/http/impl/HttpClientRequestImpl.java (hunk 2 above)
Comment (cescoffier) [inline src/main/java/io/vertx/core/http/impl/HttpClientRequestImpl.java]: Ok, but then all access to the field must be guarded. It's not the case. So there are race conditions.

**Comment by purplefox** [comment]
A different (and safer) way of implementing this, which does not involve changing the logic of HttpClientRequestImpl at all is simply to make a normal request and do the redirect logic in the standard response handler using the public API, this could be hidden in a helper method in HttpClientImpl.

I think I would prefer that as it wouldn't require fiddling with the internals of HttpClientImpl and reasoning about thread safety.

**Comment by cescoffier** [comment]
I also agree, it would be safer.

**Comment by luengnat** [comment]
It doesn't change the internal state of httpclientimpl. It only deal with the transition from one httpclientrequestimpl to the new one after the redirect. If all of these operations happen in the same event loop thread, it should be thread safe right?

**Comment by purplefox** [comment]
Closing this, but I am still open to the idea of this functionality, just not implemented this way.

**Pull request closed without merging**